tracer/src/Datadog.Trace/Logging/LogsInjection/LibLogScopeEventSubscriber.cs

```diff
@@ -168,37 +161,57 @@ public void OnTraceStarted_RefreshIISState(object sender, SpanEventArgs spanEven
         }
 #endif
 
-        public void StackOnSpanOpened(object sender, SpanEventArgs spanEventArgs)
+        public void MapOnSpanOpened(object sender, SpanEventArgs spanEventArgs)
```

Comment:
This logic is really borked and is the last piece to the puzzle. Any suggestions would be welcome.

Here's the situation with NLog: Each time we "push" a context, NLog does a Dictionary set. Each time we "pop" a context, it does a Dictionary remove. In other words, it does not restore the previous state on a "pop", it just clears the key-value pair.

I started to work with the OnSpanOpened and OnSpanClosed events because this means we only modify the log context twice per span. However, as you can kind of see from the current state of this file, the logic is a mess and I haven't gotten it right yet. Most of the complications arise when I try to make sure when a span is closed out of order that it doesn't affect the log context, so I try to do a sort of backtrack to make sure the close matches.

Now that I'm typing this up at my EOD, I think I might be able to just use the SpanDeactivated event to remove the state from the log context and use the SpanActivated event to set the new context (will need to modify it to use the DistributedSpanContext).

If you peruse the code and have any suggestions, feel free to let me know 😄

Reply:
Looks like I got it working with the SpanActivated approach. I guess I just needed to get my head out of the code for a little bit 😅